scripts/new-dataset/annotated_doc.py

```diff
@@ -0,0 +1,529 @@
+        bin_rels = BinRel.format_bin_rels_dygiepp(self.bin_rels, sent_idx_tups)
+        if len(self.equiv_rels) > 0 and coref: # Some datasets don't have coreferences
+            corefs = EquivRel.format_coreferences_dygiepp(self.equiv_rels)
+        if len(self.events) > 0: # Some datasets don't have events 
+            events = Event.format_events_dygiepp(self.events, sent_idx_tups)
+
+        # Make dict
+        res = {"doc_key": self.doc_key,
+               "dataset": self.dataset,
+               "sentences": self.sents,
+               "ner": ner,
+               "relations": bin_rels}
+
+        if len(self.equiv_rels) > 0 and coref: # Some datasets don't have coreferences
+            res["clusters"] = corefs
+        if len(self.events) > 0: # Some datasets don't have events 
+            res["events"] = events
+
+        return res 
+
+
+class Ent:
+
+    def __init__(self, line):
+        """
+        Does not account for discontinuous annotations, these should have
+        been removed before creating entity objects.
+        """
+        self.ID = line[0]
+        self.label = line[1]
+        
+        # Since disjoint entities have been dropped, start and end indices will
+        # always be at the same indices in the list 
+
+        self.start = int(line[2])
+        self.end = int(line[3])
+        self.text = " ".join(line[4:])
+
+
+    def set_start_end(self, start, end):
+        """
+        Set the start_end_tups attribute. To be used to change out character
+        indices for token indices.
+
+        parameters:
+            start_end_tups, list of tuples: list of start and end token indices
+        
+        returns: None
+        """
+        self.start = start
+        self.end = end 
+        
+
+    @staticmethod
+    def char_to_token(ent_list, sentences, nlp):
+        """
+        Does the heavy lifting for converting brat format to dygiepp format.
+        Replaces the start and end attributes for entities with their corresponding 
+        token indices. Raises a warning if no alignment can be found for an entity,
+        as the entity will be dropped.
+        
+        I would appreciate feedback on this alignment approach! 
+
+        My initial thought was to do a cumulative sum of the lengths of the
+        tokens in the list in order to convert char indices to token indices.
+        However, because there usually aren't spaces before punctuation like 
+        periods and commas, but there can be before hyphens, I wasn't sure how 
+        to make sure that I didn't make an asusmption about where to insert
+        spaces when doing the cumulative sum, because even one mismatch (e.g.
+        if I assumed there shouldn't be a space between a hyphen and the 
+        last token, but then a text uses a hypen as an en dash) would cause
+        all of the rest of the entities in the doc to be thrown away. 
+
+        So the approach I settled on was to order the entities by their 
```

Comment:
If what you'd done works, seems fine to leave it. An alternative approach would be to rely on Spacy to do the indexing for you. When you process a document with Spacy, it computes start and end indices for each token, with respect to both the whole document and the sentence that the token is part of. So, if you know the character start and end indices of the entity you're looking for, you could just search for the Spacy tokens that match.

```
nlp = spacy.load("en_core_web_sm")
doc = nlp("Here's one sentence. And another sentence.")
# Skip to the last sentence.
for sent in doc.sents: pass
# Print a single word in the second sentence
print(sent[1])
# Print start character index of word in document
print(sent[1].idx)
# Print start token index of word in document
print(sent[1].i)
# Print start token index of word in sentence
print(sent[1].sent_start)
```

Reply:
Thanks for the feedback! I changed it to use spacy, as that felt safer/more robust.